### PR TITLE
[Bugfix][Structured Output] Stage V2 grammar bitmask copies through preallocated pinned buffers

### PR DESCRIPTION
**Pull request by anencore94**
## Purpose

`StructuredOutputsWorker.apply_grammar_bitmask` (V2 runner) allocates pinned host memory **twice on every decode step** of every batch that contains structured-output requests:

- `async_copy_to_gpu(grammar_bitmask, ...)` calls `x.pin_memory()`. `grammar_bitmask` is a fresh `np.ndarray` built by the scheduler each step, so the `# pin_memory() is no-op if the memory is already pinned` assumption never holds on this path — it is a real allocation plus copy per step.
- the bitmask → logits mapping is materialized with `torch.tensor(mapping, ..., pin_memory=True)`.

Allocating pinned memory serializes on the CUDA driver's host allocator, so under concurrency the cost lands on the engine loop rather than the GPU (GPU sits mostly idle while inter-token latency inflates).

This is the **V2-runner counterpart** of the regression reported in #49013 and bisected there to #45424. The two open PRs for that issue (#49150, #49168) only change `vllm/v1/structured_output/utils.py`, i.e. the V1 path, so dense models — which move V1 → V2 across that version window — are not covered by either.

Note also that on this path the `pin_memory()` call reinstates exactly what #37006 (*"[V1] Remove pin_memory() in async_copy_to_gpu to fix sporadic stalls"*, merged) had deliberately removed, together with its comment:

```python
# Copy directly to GPU — explicit pin_memory() causes sporadic stalls
# under high concurrency due to CUDA driver contention. The driver
# handles the transfer efficiently without manual pinning.
```

## Changes

Stage both copies through `CpuGpuBuffer`s sized to `max_num_logits`, allocated once in `__init__`. This keeps #45424's property that the source of an async h2d copy is pinned, while removing the per-step allocation. `CpuGpuBuffer` is already used the same way elsewhere in the V2 worker (e.g. `model_states/mamba_hybrid.py`).

Buffer reuse follows the same pattern as the rest of the V2 input pipeline: host writes for step N+1 are ordered after step N's copies by the existing sampler synchronization, and only the rows in use (`[:num_masks]`, `[:num_logits]`) are written and copied.

## Test Result

Measured on 1×B300, `google/gemma-4-12B-it`, 172-core host with a 12-CPU cgroup quota, 80 requests at concurrency 8, every request guided-JSON (xgrammar), identical args otherwise:

| build | guided p50 | decode throughput |
|---|---|---|
| v0.25.1 | 7.49 s | 115 tok/s |
| v0.25.1 + this patch | **1.07 s** | **740 tok/s** |
| v0.25.1 + this patch (2nd run, warm) | 1.06 s | 806 tok/s |

Guided decoding returns to parity with unguided on the same build (1.07 s vs 1.11 s p50). Unguided throughput is unchanged, as expected — `apply_grammar_bitmask` returns early on `if not grammar_req_ids`.

Verified by mounting the patched file into an otherwise unmodified `vllm/vllm-openai:v0.25.1` container, so the only delta is this diff.

Added `tests/v1/worker/test_structured_outputs_staging.py` covering: staging buffers are preallocated at the batch upper bound, host/device pointers are stable across steps, the staged bitmask masks exactly the disallowed token ids, stale rows from a wider previous step do not leak, and the no-request path is a no-op.

## Related

- #49013 — the umbrella perf issue (V1 path bisect)
- #49150, #49168 — V1-path fixes; this PR is complementary, not overlapping
- #45424 — the commit that introduced the per-step pinning
- #37006 — the earlier fix this silently reverted on the V2 path

### CODE REVIEW

**Comment by njhill** [comment]
Thanks @anencore94. Please try https://github.com/vllm-project/vllm/pull/49919 which should hopefully fix the performance regression.

**Comment by njhill** [comment]
@anencore94 closing now that https://github.com/vllm-project/vllm/pull/49919 is merged. Please speak up if it didn't fix the problem for you.
